akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -44,87 +45,75 @@ class FoldWithin[In, Agg, Out](
-        // although we only need this timer when the aggregator is not empty
-        // and this event do not need to be triggered otherwise,
-        // it's more efficient to set up a upfront timer
-        maxGap.foreach(mg => scheduleWithFixedDelay(maxGapTimer, mg, mg))
+        // although we do not need timer events when the group is empty
+        // it's more efficient to set up a upfront timer with fixed interval
+        // as opposed to time every single event
+        // always pick the smaller of the two intervals and divide by 2
+        Option(maxGap.getOrElse(maxDuration.getOrElse(null))).map(_ / 2).map {
+          delay => scheduleWithFixedDelay("FoldWithinIntervalTimer", delay, delay)
+        }
       }
-      override protected def onTimer(timerKey: Any): Unit = state.harvestOnTimer()
+
+      override protected def onTimer(timerKey: Any): Unit = if (readyToEmitByTimeout) harvestAndEmit()
 
       override def onPush(): Unit = {
-        state.aggregate()
+        aggregateAndEmitIfReady()
         if (isAvailable(out)) pull(in) // pull only if there is no backpressure from outlet
       }
 
       override def onUpstreamFinish(): Unit = {
-        state.harvestOnFlush()
+        harvestAndEmit()
         completeStage()
       }
 
       override def onPull(): Unit = {
-        state.aggregate()
+        aggregateAndEmitIfReady()
         if (!hasBeenPulled(in)) pull(in) // always pass through the pull to upstream
       }
 
       setHandlers(in, out, this)
 
-      // mutable state to keep track of the aggregator status to coordinate the flow
-      // input/output handler callbacks are guaranteed to execute without concurrency
-      // https://doc.akka.io/docs/akka/current/stream/stream-customize.html#thread-safety-of-custom-operators
-      private val state = new State()
-
-      class State {
-
-        private var aggregator: Option[AggregatorState] = None
-
-        def harvestOnTimer(): Unit = harvestWithMode(mode = HarvestMode.OnTimer)
-
-        def harvestOnFlush(): Unit = harvestWithMode(mode = HarvestMode.Flush)
-
-        private def harvestWithMode(mode: HarvestMode.Value): Unit = {
-          aggregator foreach {
-            case agg =>
-              if (mode == HarvestMode.Flush || emitReady(agg.aggregator) || {
-                mode == HarvestMode.OnTimer && {
-                  val currentTime = System.nanoTime()
-                  // check gap only on timer
-                  maxGap.exists(mg => currentTime - agg.lastAggregateTime >= mg.toNanos) ||
-                    maxDuration.exists(md => currentTime - agg.startTime >= md.toNanos)
-                }
-              }) {
-                aggregator = None
-                emit(out, FoldWithin.this.harvest(agg.aggregator)) // if out port not available, it'll follow up as scheduled emit
-              }
-          }
-        }
+      private def harvestAndEmit(): Unit = {
+        aggregator.toOption.foreach(
```

Comment:
@johanandren @patriknw please let me know if there is obvious concerns of the implementation I'll work on testing.

Reply:
looking good

Reply:
Please go ahead and work on tests. 

One corner case I think would be worth to verify with tests is if the timeout hits to emit a group but downstream is back pressuring, so it can't, and then enough time passes so that it still is applying back pressure when the next timeout hits.

Reply:
test cases added using simulated time